Add ``prng`` parameter to randMatrix

urkud opened the pull request:
This can be useful, e.g., to generate a Matrix with known seed, then use the
new state of prng to generate other objects.

urkud committed:
Add ``prng`` parameter to randMatrix

This can be useful, e.g., to generate a Matrix with known seed, then use the
new state of prng to generate other objects.

asmeurer commented on sympy/matrices/dense.py:
```diff
@@ -1570,12 +1570,21 @@ def casoratian(seqs, n, zero=True):
+    The pseudo-random number generator used to generate matrix is chosen in the
+    following way:
+
+    * if ``prng`` is supplied, it will be used as random number generator;
```
You should specify that it has to be a `random.Random` class (or at least have the same API).

toolforger replied:
Maybe "same API" is too disheartening, it's a huge API.
"`random.Random` or a subclass" would work, and would be what anybody in his right mind should do when writing a compatible class.

asmeurer commented:
I like this pull request, but I don't like the option name `prng`. `random_class` is more descriptive.

asmeurer commented on sympy/matrices/dense.py:
```diff
@@ -1570,12 +1570,21 @@ def casoratian(seqs, n, zero=True):
+    * if ``prng`` is supplied, it will be used as random number generator;
+    * if ``prng`` is not supplied but ``seed`` is supplied, then new
+      :class:`random.Random` with given ``seed`` will be created;
+    * otherwise, a new :class:`random.Random` with default seed (system time)
```
The system time thing is not always true. I would just remove it.

toolforger commented on sympy/matrices/dense.py:
```diff
@@ -1570,12 +1570,21 @@ def casoratian(seqs, n, zero=True):
 
 
-def randMatrix(r, c=None, min=0, max=99, seed=None, symmetric=False, percent=100):
+def randMatrix(r, c=None, min=0, max=99, seed=None, prng=None, symmetric=False, percent=100):
```
The new parameter should be added last, in the (admittedly rare) case that somebody provided parameters by position.

toolforger commented:
> I don't like the option name `prng`.

That's IT lingo for "pseudo random number generator".

 > `random_class` is more descriptive.

The parameter is used as a generator object, not as a class, so 
`xxx_class` would be quite misleading.
I can't think of a good name that isn't overly long. Maybe somebody else 
can; otherwise, I'd recommend sticking with `prng` and add a link to 
https://en.wikipedia.org/wiki/PRNG in the docstring.

asmeurer commented:
The parameter is used as a class. Look at the code.

toolforger commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -480,12 +481,19 @@ def test_expand():
     assert S == M  # doesn't fail when elements are Numbers, not int
 
+    rng = random.Random(4)
+    assert M == randMatrix(3, prng=rng, symmetric=True)
```
You can use `prng` instead of `rng` here.
Or just write `assert M == randMatrix(3, prng=random.Random(4), symmetric=True)`.

toolforger commented:
> The parameter is used as a class. Look at the code.

The parameter is used as an instance. Look at the code:

`prng = prng or random.Random(seed)`

Assigns a freshly created `Random` object to `prng` if `prng` is `None`.

`        m = Matrix._new(r, c, lambda i, j: prng.randint(min, max))`

Calls `prng.randint`. 
https://hg.python.org/cpython/file/2.7/Lib/random.py says that `randint` 
is not a class function.
(Equivalent for `shuffle`.)

asmeurer commented:
Since this PR is from Jan 5, I don't know if @urkud will finish it. Someone may want to pick up on the branch and start a new PR that addresses these issues.

urkud commented:
Hi!

I can finish it, but I have no access to the computer that remembers my github credentials.
I'll have some time the day after tomorrow. It seems that all issues are trivial, so it shouldn't take much time to fix them.

On 13 July 2015 20:37:34 GMT+03:00, Aaron Meurer <notifications@github.com> wrote:
>Since this PR is from Jan 5, I don't know if @urkud will finish it.
>Someone may want to pick up on the branch and start a new PR that
>addresses these issues. 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/sympy/sympy/pull/8752#issuecomment-121001492

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

asmeurer commented:
You can add your GitHub ssh keys to any computer https://help.github.com/articles/generating-ssh-keys/

urkud commented:
Moved prng parameter to the end, fixed docstring, rebased. Any other issues?

toolforger commented on sympy/matrices/dense.py:
```diff
@@ -1617,10 +1628,8 @@ def randMatrix(r, c=None, min=0, max=99, seed=None, symmetric=False, percent=100
-        prng = random.Random()  # use system time
-    else:
-        prng = random.Random(seed)
+    # Note that ``Random()`` is equivalent to ``Random(None)``
```
Documenting that not handling `None` explicitly is a good idea.

toolforger commented:
All points that I raised are addressed.
